code/datums/components/manual_heart.dm

```diff
@@ -0,0 +1,170 @@
+/**
+ * Manual heart pumping component. Requires the holder to pump their heart manually every
+ * so often or die.
+ *
+ * Mainly used by the cursed heart.
+ */
+/datum/component/manual_heart
+	/// The action for pumping your heart
+	var/datum/action/cooldown/manual_heart/pump_action
+	var/last_pump = 0
+	var/add_colour = TRUE //So we're not constantly recreating colour datums
+	/// How long between needed pumps; you can pump one second early
+	var/pump_delay = 3 SECONDS
+	/// How much blood volume you lose every missed pump, this is a flat amount not a percentage!
+	var/blood_loss = BLOOD_VOLUME_NORMAL * 0.2 // 20% of normal volume, missing five pumps is instant death
+
+	//How much to heal per pump, negative numbers would HURT the player
+	var/heal_brute = 0
+	var/heal_burn = 0
+	var/heal_oxy = 0
+
+/datum/component/manual_heart/Initialize(pump_delay = 3 SECONDS, blood_loss = BLOOD_VOLUME_NORMAL * 0.2, heal_brute = 0, heal_burn = 0, heal_oxy = 0)
+	//Non-Carbon mobs can't have hearts, and should never receive this component.
+	if (!iscarbon(parent))
+		stack_trace("Manual Heart component added to [parent] ([parent?.type]) which is not a /mob/living/carbon subtype.")
+		return COMPONENT_INCOMPATIBLE
+
+	src.pump_delay = pump_delay
+	src.blood_loss = blood_loss
+	src.heal_brute = heal_brute
+	src.heal_burn = heal_burn
+	src.heal_oxy = heal_oxy
+
+	pump_action = new(src)
+	pump_action.cooldown_time = pump_delay - (1 SECONDS) //you can pump up to a second early
+	pump_action.Grant(parent)
+
+	var/mob/living/carbon/carbon_parent = parent
+	var/obj/item/organ/internal/heart/parent_heart = carbon_parent.get_organ_slot(ORGAN_SLOT_HEART)
+	if(parent_heart && !HAS_TRAIT(carbon_parent, TRAIT_NOBLOOD) && carbon_parent.stat != DEAD)
+		START_PROCESSING(SSdcs, src)
+		last_pump = world.time
+
+	to_chat(parent, span_userdanger("Your heart no longer beats automatically! You have to pump it manually - otherwise you'll die!"))
+
+/datum/component/manual_heart/Destroy()
+	to_chat(parent, span_userdanger("You feel your heart start beating normally again!"))
+	var/mob/living/carbon/carbon_parent = parent
+	if(istype(carbon_parent))
+		carbon_parent.remove_client_colour(/datum/client_colour/manual_heart_blood)
```

Comment:
Doesn't make a huge amount of difference but I'd move this to `UnregisterFromParent`